lib/Mojolicious/Plugin/AssetPack/Pipe/OptiPng.pm

```diff
@@ -0,0 +1,94 @@
+
+package Mojolicious::Plugin::AssetPack::Pipe::OptiPng;
+
+use Mojo::Base 'Mojolicious::Plugin::AssetPack::Pipe';
+use Mojo::Util qw( spurt slurp );
+use File::Temp qw( tempfile );
+
+sub process {
+	my ($self, $assets) = @_;
+	my $store = $self->assetpack->store;
+	my $file;
+
+	$assets->each(
+		sub {
+			my ($asset, $index) = @_;
+
+			# might also use optipng to convert BMP, GIF, PNM or TIFF to PNG
+			# but not yet
+			return unless $asset->format eq 'png';
+
+			my $attrs = $asset->TO_JSON;
+			$attrs->{key}    = 'optipng';
+			$attrs->{format} = 'png';
+
+			# already processed before?
+			return $asset->content($file)->FROM_JSON($attrs) if $file = $store->load($attrs);
+
+			my ( $content, $error );
+			{
+				my ( $fh, $file ) = tempfile();
+				spurt $store->load($asset), $file;
+
+				my ( $in, $out, $err ) = ( '', '', '' );
+				my @cmd = ( 'optipng', '-quiet', '-clobber', $file );
```

Comment:
Not sure why you talk about STDIN/STDOUT. What does that have to do with run()? IPC::Run3 has support for reading/writing STDIN/STDOUT/STDERR, but that doesn't mean you need to use those features.

run() (as documented) also provides setting ```$MOJO_ASSETPACK_OPTIPNG_APP``` in case you don't have the application in $PATH.

Maybe the most important part about run() is that a ```sub _install_optipng {}``` method can provide information about how to install optipng or simply install it (if possible).